### PR TITLE
add typescript defs for new features (close #2428)

### LINKED ISSUE #2428: Type definitions for Request Hooks
### Are you requesting a feature or reporting a bug?
I'm requesting a feature.

### What is the current behavior?
Typescript compilation fails when trying to use Request Hooks.

### What is the expected behavior?
It is possible to write tests in Typescript and use request hooks.

### How would you reproduce the current behavior (if this is a bug)?

```ts
import { RequestMock } from 'testcafe'; // Error: Module 'testcafe' has no exported member 'RequestMock'.

const mock = RequestMock()
  .onRequestTo('')
  .respond('', 200);

fixture(`Test page`).page(`http://127.0.0.1:4200`);

test.only('Request Hooks', async t => {
  await t.addRequestHooks(mock); // Error: Property 'addRequestHooks' does not exist on type 'TestController'.
});

```

### Specify your

* testcafe version: 0.20.0
* node.js version: 10.0.0
* typescript version: 2.7.2

### PR DESCRIPTION
**Pull request by miherlosev**
@AndreyBelym @MargaritaLoseva 

Changes:
* change var to const in header of `test/server/compiler-test.js`
* `resolve` -> `path.resolve`
* rewrite `Should provide API definitions` using `fs.readDirSync`
* add typescript definitions for `meta` and `Request Hook` features.

### CODE REVIEW

File: ts-defs/index.d.ts [hunk 1]
```diff
@@ -863,14 +863,130 @@ interface ResizeToFitDeviceOptions {
...
+    respond(body?: object | string | ((req, res) => any), statusCode?: number, headers?: object): RequestMock;
+}
+
+interface Request {
+    /**
+     * The user agent that sent the request.
+     */
+    userAgent: string;
+    /**
+     * Request part of the logged request
```
Comment (MargaritaLoseva) [inline ts-defs/index.d.ts]: The request part of the logged request

File: ts-defs/index.d.ts [hunk 2]
```diff
@@ -863,14 +863,130 @@ interface ResizeToFitDeviceOptions {
...
+     */
+    request: RequestData;
+    /**
+     * Response part of the logged request
```
Comment (MargaritaLoseva) [inline ts-defs/index.d.ts]: The response part of the logged request

File: ts-defs/index.d.ts [hunk 3]
```diff
@@ -1151,6 +1267,18 @@ interface TestController {
      * @param role - The role you need to use further in the test.
      */
     useRole(role: Role): TestControllerPromise;
+    /**
+     * Attach the hooks during a test run
```
Comment (MargaritaLoseva) [inline ts-defs/index.d.ts]: Attaches

File: ts-defs/index.d.ts [hunk 4]
```diff
@@ -1434,6 +1562,37 @@ declare module 'testcafe' {
      */
     export function ClientFunction(fn: Function, options?: ClientFunctionOptions): ClientFunction;
 
+    /**
+     * Creates a RequestMock
```
Comment (MargaritaLoseva) [inline ts-defs/index.d.ts]: Creates RequestMock

File: ts-defs/index.d.ts [hunk 5]
```diff
@@ -1434,6 +1562,37 @@ declare module 'testcafe' {
...
+     */
+    export function RequestMock(): RequestMock;
+
+    /**
+     * Creates a RequestLogger
```
Comment (MargaritaLoseva) [inline ts-defs/index.d.ts]: Creates RequestLogger

File: ts-defs/index.d.ts [hunk 6]
```diff
@@ -1434,6 +1562,37 @@ declare module 'testcafe' {
...
+     */
+    export function RequestLogger(filter?: string | RegExp | object | ((req, res) => boolean), options?: RequestLoggerOptions): RequestLogger;
+
+    /** The RequestHook class used to create a custom HTTP Request Hook **/
```
Comment (MargaritaLoseva) [inline ts-defs/index.d.ts]: The RequestHook class used to create a custom HTTP request hook

File: ts-defs/index.d.ts [hunk 7]
```diff
@@ -1434,6 +1562,37 @@ declare module 'testcafe' {
...
+    export class RequestHook {
+        /**
+         * Creates a Request Hook
```
Comment (MargaritaLoseva) [inline ts-defs/index.d.ts]: Creates a request hook

File: ts-defs/index.d.ts [hunk 8]
```diff
@@ -1491,7 +1650,7 @@ interface FixtureFn {
      * @param url - The URL of the webpage at this tests start.
```
Comment (MargaritaLoseva) [inline ts-defs/index.d.ts]: The URL of the webpage where tests start.

File: ts-defs/index.d.ts [hunk 9]
```diff
@@ -1534,6 +1693,25 @@ interface FixtureFn {
      * Skips execution of all tests, except whose that are in this fixture.
      */
     only: this;
+    /**
+     * Specifies the additional information for all tests in the fixture that can be used in reports
```
Comment (MargaritaLoseva) [inline ts-defs/index.d.ts]: Specifies the additional information for all tests in the fixture. This information can be used in reports.

File: ts-defs/index.d.ts [hunk 10]
```diff
@@ -1534,6 +1693,25 @@ interface FixtureFn {
...
+     *
+     * @param key - The name of the metadata entry
+     * @param value - The value of the metadata entry
+     */
+    meta(key: string, value: string): this;
+    /**
+     * Specifies the additional information for all tests in the fixture that can be used in reports
```
Comment (MargaritaLoseva) [inline ts-defs/index.d.ts]: Specifies the additional information for all tests in the fixture. This information can be used in reports.

File: ts-defs/index.d.ts [hunk 11]
```diff
@@ -1534,6 +1693,25 @@ interface FixtureFn {
...
+     *
+     * @param data - Key-value pairs
+     */
+    meta(data: object): this;
+    /**
+     * Attaches hooks to all tests in the fixture
+     *
+     * @param hook - The set of RequestHook subclasses
```
Comment (MargaritaLoseva) [inline ts-defs/index.d.ts]: The set of the RequestHook subclasses

File: ts-defs/index.d.ts [hunk 12]
```diff
@@ -1580,6 +1758,25 @@ interface TestFn {
      * Skips execution of all tests, except this one.
      */
     only: this;
+    /**
+     * Specifies the additional information for the test that can be used in reports
```
Comment (MargaritaLoseva) [inline ts-defs/index.d.ts]: Specifies the additional information for the test. This information can be used in reports.

File: ts-defs/index.d.ts [hunk 13]
```diff
@@ -1580,6 +1758,25 @@ interface TestFn {
...
+     *
+     * @param key - The name of the metadata entry
+     * @param value - The value of the metadata entry
+     */
+    meta(key: string, value: string): this;
+    /**
+     * Specifies the additional information for the test that can be used in reports
```
Comment (MargaritaLoseva) [inline ts-defs/index.d.ts]: Specifies the additional information for the test. This information can be used in reports.

File: ts-defs/index.d.ts [hunk 14]
```diff
@@ -1580,6 +1758,25 @@ interface TestFn {
...
+     *
+     * @param data - Key-value pairs
+     */
+    meta(data: object): this;
+    /**
+     * Attaches hooks to the test
+     *
+     * @param hook - The set of RequestHook subclasses
```
Comment (MargaritaLoseva) [inline ts-defs/index.d.ts]: The set of the RequestHook subclasses

**Comment by miherlosev** [comment]
@testcafe-build-bot \retest

**Comment by miherlosev** [comment]
@testcafe-build-bot \retest

**Comment by miherlosev** [comment]
@MargaritaLoseva review please

**Comment by miherlosev** [comment]
@testcafe-build-bot \retest

**Comment by miherlosev** [comment]
@testcafe-build-bot \retest

**Pull request merged**